contracts/staking/fuse/utils/StakingRewardsPerEpoch.sol

```diff
@@ -0,0 +1,201 @@
+		nonReentrant
+		updateReward(msg.sender)
+	{
+		require(_amount > 0, "Cannot withdraw 0");
+		_withdraw(msg.sender, _amount);
+	}
+
+	function getReward() public nonReentrant updateReward(msg.sender) {
+		_getReward(msg.sender);
+	}
+
+	function exit() external {
+		withdraw(stakersInfo[msg.sender].balance);
+		getReward();
+	}
+
+	function rewardPerToken() public view returns (uint256) {
+		if (_totalSupply == 0) {
+			return rewardPerTokenAt[0];
+		}
+		return
+			rewardPerTokenStored +
+			((lastTimeRewardApplicable() - lastUpdateTime) * rewardRate * PRECISION) /
+			_totalSupply;
+	}
+
+	function _getRewardPerTokenPerUser(address _account)
+		internal
+		view
+		returns (uint256)
+	{
+		return
+			rewardsPerTokenAt[lastEpochIndex] -
+			rewardsPerTokenAt[stakersInfo[_account].indexOfLastEpochStaked + 1];
+	}
+
+	function earned(address account) public view  returns (uint256) {
+		return
+			(stakersInfo[account].balance *
+				(rewardPerToken() -
+					stakersInfo[account].rewardPerTokenPaid)) /
+			PRECISION +
+			stakersInfo[account].reward;
+	}
+
+	function _updateReward(address _account) internal virtual {
+		lastUpdateTime = lastTimeRewardApplicable();
+		if (_account != address(0)) {
+			_addPendingStakesToBalanceOnTimeUpdate(_account);
+			stakersInfo[_account].reward = earned(_account);
+			stakersInfo[_account]
+					.rewardPerTokenPaid = _getRewardPerTokenPerUser(_account);
+		}
+	}
+
+	function _addPendingStakesToBalanceOnTimeUpdate(address _account) internal {
+		if (
+					stakersInfo[_account].indexOfLastEpochStaked !=
+					lastEpochIndex
+		) {
+				stakersInfo[_account].balance += stakersInfo[_account]
+						.pendingStake;
+				stakersInfo[_account].pendingStake = 0;
+		}
+	}
+
+	function _updatePeriodAndRate(uint256 reward) internal virtual {
+		if (block.timestamp >= periodFinish) {
+			rewardRate = reward / rewardsDuration;
+		} else {
+			uint256 remaining = periodFinish - block.timestamp;
+			uint256 leftover = remaining * rewardRate;
+			rewardRate = (reward + leftover) / rewardsDuration;
+		}
+
+		// Ensure the provided reward amount is not more than the balance in the contract.
+		// This keeps the reward rate in the right range, preventing overflows due to
+		// very high values of rewardRate in the earned and rewardsPerToken functions;
+		// Reward + leftover must be less than 2^256 / 10^18 to avoid overflow.
+		require(
+			rewardRate <= rewardsToken.balanceOf(address(this)) / rewardsDuration,
+			"Provided reward too high"
+		);
+
+		lastUpdateTime = block.timestamp;
+		periodFinish = block.timestamp + rewardsDuration;
+		emit RewardAdded(reward);
+	}
+
+	function _notifyRewardAmount(uint256 reward)
+		internal
+		updateReward(address(0))
+	{
+		_totalSupply += pendingStakes;
+		rewardsPerTokenAt.push(rewardPerToken());
+		lastEpochIndex++;
+		_updatePeriodAndRate(reward);
+	}
+
+	function _withdraw(address _from, uint256 _amount) internal virtual {
+		if (stakersInfo[_from].pendingStake >= _amount) {
```

Comment:
what about the `else` case? (pendingtake>0 and < _amount)
there should always be something happening `if pendingStake>0` that's the correct initial if

Reply:
Then it'll fail in safeTransfer.

Reply:
Why we have to do something if the pendingStake == 0 or < _amount. Just let it crash in transfer, so user would have to use the correct amount.

Reply:
If pendingStake>0 and < _amount we should subtract the pendingStake that's true. Added the change.
e.g. staker staked 100 in the past, in current epoch 20 that are pending, so if he withdraws 50 we should sub 20.